### PR TITLE
#2604 rebased - fork() based garbage collector

### PR DESCRIPTION
**Pull request by rainers**
Attempt to rebase https://github.com/dlang/druntime/pull/2604 to see whether it still fails

### CODE REVIEW

**Comment by RazvanN7** [comment]
Thanks for doing this @rainers ! Fingers crossed! Things look good momentarily.

**Comment by ibuclaw** [comment]
https://sourceware.org/bugzilla/show_bug.cgi?id=4737

On 2.34 with now have _Fork (c32c868ab8b2b) which is a async-signal-safe fork()
replacement added by Austin Group issue 62 [1].  The Austin defect also dropped
the async-signal-safe requirement for fork, so currently there is no plan to
try make fork() async-signal-safe.

**Comment by rainers** [comment]
@ibuclaw Thanks for the link. It seems the fork issue might actually be causing the failures (most seem to be deadlocking and are terminated by timeout).

libc 2.34 doesn't seem to be released yet and might take some time before being available in major distros. Do you think it would be possible now to emulate what `_Fork` does, i.e. do the appropriate syscall (at least on linux)?

**Comment by ibuclaw** [comment]
> @ibuclaw Thanks for the link. It seems the fork issue might actually be causing the failures (most seem to be deadlocking and are terminated by timeout).
> 
> libc 2.34 doesn't seem to be released yet and might take some time before being available in major distros. Do you think it would be possible now to emulate what `_Fork` does, i.e. do the appropriate syscall (at least on linux)?

Possibly https://github.com/bminor/glibc/blob/08cbcd4dbc686bb38ec3093aff2f919fbff5ec17/sysdeps/unix/sysv/linux/arch-fork.h

**Comment by rainers** [comment]
> Possibly https://github.com/bminor/glibc/blob/08cbcd4dbc686bb38ec3093aff2f919fbff5ec17/sysdeps/unix/sysv/linux/arch-fork.h

I have tried to imitate that for Linux/x86_64. Let's see if this platform passes now.

**Comment by rainers** [comment]
@FraMecca Hi. In an attempt to resurrect your efforts I rebased your PR and squashed your commits here. I hope that is ok with you. Should I rather push to your PR #2604 instead?

**Comment by rainers** [comment]
This has passed all tests with the forking GC enabled, so should be ready for review. 

- the forking was only enabled by default to use it in all tests, should be disabled by default.
- parallel marking is disabled when forking. In principle it might be interesting to do this in the forked process, but that means the additional threads have to be started every time - not optimal. In addition they would have to be limited somehow as using all CPUs for marking defeats the purpose of the concurrent GC, i.e. keeping the application running.

Running the GC benchmarks (in a VM with 4 cores on an i5-2865U), I get
```
MIN conalloc         1.557 s,   153 MB,   24 GC  148 ms, Pauses   39 ms <    2 ms
MIN conappend        0.977 s,    91 MB,   34 GC   78 ms, Pauses   38 ms <    2 ms
MIN concpu           1.458 s,   181 MB,   23 GC  132 ms, Pauses   34 ms <    3 ms
MIN conmsg           4.930 s,   176 MB,    7 GC   21 ms, Pauses   11 ms <    2 ms
MIN dlist            6.088 s,   145 MB,    8 GC   19 ms, Pauses    7 ms <    1 ms
MIN huge_single      0.007 s,  1504 MB,    2 GC    0 ms, Pauses    0 ms <    0 ms
MIN rand_large       4.610 s,  1113 MB, 1751 GC  865 ms, Pauses  238 ms <    0 ms
MIN rand_small      24.385 s,  2443 MB,   46 GC 1018 ms, Pauses  481 ms <   23 ms
MIN slist            6.173 s,   145 MB,    8 GC   18 ms, Pauses    7 ms <    1 ms
MIN testgc3          3.196 s,   425 MB,    5 GC   19 ms, Pauses    7 ms <    5 ms
MIN tree1            1.382 s,   117 MB,    8 GC    9 ms, Pauses    5 ms <    1 ms
MIN tree2            4.362 s,    92 MB,    7 GC    6 ms, Pauses    3 ms <    0 ms
MIN vdparser         7.434 s,   145 MB,    8 GC   23 ms, Pauses    7 ms <    1 ms
MIN words            3.303 s,   386 MB,    3 GC    1 ms, Pauses    0 ms <    0 ms
```
The last value on each line is the most interesting one: the maximum pause time. Compare this with the non-forking GC (with parallel marking disabled):
```
MIN conalloc         1.657 s,    12 MB, 1085 GC  448 ms, Pauses  442 ms <    1 ms
MIN conappend        0.857 s,    12 MB,  707 GC  281 ms, Pauses  276 ms <    2 ms
MIN concpu           2.033 s,    12 MB, 1166 GC  808 ms, Pauses  803 ms <    3 ms
MIN conmsg           4.088 s,    51 MB,   31 GC   96 ms, Pauses   90 ms <    5 ms
MIN dlist            5.792 s,    22 MB,   35 GC  160 ms, Pauses  158 ms <    5 ms
MIN huge_single      0.008 s,  1504 MB,    2 GC    0 ms, Pauses    0 ms <    0 ms
MIN rand_large       5.346 s,   171 MB, 8027 GC 2143 ms, Pauses 1690 ms <    1 ms
MIN rand_small      11.547 s,    12 MB, 7637 GC 2318 ms, Pauses 1924 ms <    1 ms
MIN slist            5.712 s,    12 MB,  101 GC  131 ms, Pauses  128 ms <    2 ms
MIN testgc3          2.988 s,   258 MB,   10 GC  232 ms, Pauses  222 ms <   47 ms
MIN tree1            1.203 s,     5 MB,   64 GC   27 ms, Pauses   26 ms <    1 ms
MIN tree2            4.238 s,     1 MB,  216 GC   48 ms, Pauses   47 ms <    0 ms
MIN vdparser         6.746 s,    70 MB,   13 GC  121 ms, Pauses  117 ms <   17 ms
MIN words            3.157 s,   358 MB,    5 GC   11 ms, Pauses    9 ms <    8 ms
```
For reference, these are the results with parallel marking:
```
MIN conalloc         1.299 s,    12 MB,  680 GC  183 ms, Pauses  179 ms <    2 ms
MIN conappend        0.817 s,    12 MB,  746 GC  211 ms, Pauses  206 ms <    1 ms
MIN concpu           1.953 s,    12 MB, 1281 GC  675 ms, Pauses  669 ms <   12 ms
MIN conmsg           3.975 s,    22 MB,   84 GC  102 ms, Pauses   97 ms <    3 ms
MIN dlist            5.750 s,    22 MB,   35 GC  277 ms, Pauses  274 ms <   13 ms
MIN huge_single      0.007 s,  1504 MB,    2 GC    0 ms, Pauses    0 ms <    0 ms
MIN rand_large       4.614 s,   171 MB, 7858 GC 1438 ms, Pauses  952 ms <    1 ms
MIN rand_small      11.495 s,    12 MB, 7637 GC 2131 ms, Pauses 1710 ms <    0 ms
MIN slist            5.784 s,    12 MB,  101 GC  198 ms, Pauses  196 ms <    3 ms
MIN testgc3          3.031 s,   258 MB,   10 GC  223 ms, Pauses  212 ms <   37 ms
MIN tree1            1.226 s,     5 MB,   64 GC   28 ms, Pauses   25 ms <    1 ms
MIN tree2            4.158 s,     1 MB,  216 GC   30 ms, Pauses   28 ms <    0 ms
MIN vdparser         6.762 s,    70 MB,   13 GC   79 ms, Pauses   74 ms <   10 ms
MIN words            3.115 s,   358 MB,    5 GC   16 ms, Pauses   14 ms <   13 ms
```

Observations:
- the forking GC very much reduces pause time.
- it can need quite a bit more memory as allocation continues while a concurrent collection is running. Benchmarks can be pretty bad in this regard if allocating is all they do.

**Comment by kinke** [comment]
fork:
> MIN rand_small      24.385 s,  2443 MB,   46 GC 1018 ms, Pauses  481 ms <   23 ms

parallel marking:
> MIN rand_small      11.495 s,    12 MB, 7637 GC 2131 ms, Pauses 1710 ms <    0 ms

Does this mean the forking GC increases the runtime by > 100% and requires more than 200x as much RAM?!

**Comment by rainers** [comment]
> Does this mean the forking GC increases the runtime by > 100% and requires more than 200x as much RAM?!

I guess this is an extreme example made worse by running in a VM. All the test does is allocating random chunks of memory. That's a pattern that isn't handled very well as there is no throttling of the allocation rate.

GC time is quite a bit less than without forking, but overall run time is usually worse. I suspect this is the penalty for COW, which is probably exaggerated by the VM. It would be nice if someone could test this on a native linux system.

**Comment by rainers** [comment]
Hi @llucax, you might be interested in this, too.
I've (hopefully) worked around the deadlock issue by using `clone()` instead of `fork()`. At least tests are passing rather consistently for a week now.

**Comment by llucax** [comment]
> Hi @llucax, you might be interested in this, too.
> I've (hopefully) worked around the deadlock issue by using `clone()` instead of `fork()`. At least tests are passing rather consistently for a week now.

Hi @rainers, thanks! As @ibuclaw said, I recently got a notification about the new `_Fork` and automatically thought about this :)
I had the feeling there was a reason why `clone` wasn't a real solution, but now I can't remember why (I think some other interactions with libc), but if tests are running fine, I guess this is really good news!

I'll have a look at when I have some time, at least to the clone part, as the rest if it's what @FraMecca did, I already reviewed it during SAOC.

**Comment by llucax** [comment]
> I guess this is an extreme example made worse by running in a VM. All the test does is allocating random chunks of memory. That's a pattern that isn't handled very well as there is no throttling of the allocation rate.
> 
> GC time is quite a bit less than without forking, but overall run time is usually worse. I suspect this is the penalty for COW, which is probably exaggerated by the VM. It would be nice if someone could test this on a native linux system.

I want to add to this that in my experiments (that are almost 11 years old :scream: ) I notice in real programs (actually I only tested one real program) the total runtime was less, which was unexpected. The COW and other overhead should make the totl runtime worse, **but** it also adds concurrency for free, at least for single-threaded programs (so this benefits might only be present in such programs), so if instead of waiting for the GC to finish to do actual calculations you are still doing the calculations your program is supposed to do, it means it will finish earlier, and this is what I observed with this particular program (and some other benchmarks that follow this pattern, you can see in these benchmarks that `conalloc` or `concpu` have better runtimes than the non-forking GC, `concpu` is still beating the parallel marking version).

So which GC to use might depend a lot on your particular program, *there is no silver bullet*. Hopefully the default GC should be good enough for the majority of cases, but when you are bumping into walls, you certainly might benefit from trying different GC configurations.

For reference, this is the 11 years old blog post where I did the first GC benchmarks of the fully concurrent GC, `dil` is the *real* program, a D parser/compiler:
https://llucax.com/blog/blog/post/-1a4bdfba

File: src/core/internal/gc/impl/conservative/gc.d [hunk 1]
```diff
@@ -1387,6 +1387,9 @@ struct Gcx
                     auto pool = list.pool;
                     auto biti = cast(size_t)(cast(void*)list - pool.baseAddr) >> Pool.ShiftBy.Small;
                     assert(pool.freebits.test(biti));
+                    ppprev = pprev;
+                    pprev = prev;
+                    prev = list;
```
Comment (llucax) [inline src/core/internal/gc/impl/conservative/gc.d:1393]: This seems unrelated to this commit (use clone() instead of fork()).

File: src/core/internal/gc/impl/conservative/gc.d [hunk 2]
```diff
@@ -2648,29 +2668,50 @@ struct Gcx
...
+            }
+            char[256] stackbuf; // enough space to place some info for the child without stomping the parent stack
+            auto stack = stackbuf.ptr + (isStackGrowingDown ? stackbuf.length : 0);
+            auto pid = clone(&wrap_delegate, stack, flags, &dg);
+        }
+        else
+        {
+            fork_needs_lock = false;
+            auto pid = fork();
+            fork_needs_lock = true;
```
Comment (llucax) [inline src/core/internal/gc/impl/conservative/gc.d:2710]: Does it makes sense to keep supporting a fork-based approach if it's known to be buggy and lead to deadlock? Maybe look for the existence of `_Fork` instead so any OS adopting this new function can be supported too?

File: src/core/internal/gc/impl/conservative/gc.d [hunk 3]
```diff
@@ -2781,7 +2822,7 @@ Lmark:
             prepTime += (stop - start);
             start = stop;
 
-            if (doFork && !isFinal) // don't start a new fork during termination
+            if (doFork && !isFinal && !block) // don't start a new fork during termination
```
Comment (llucax) [inline src/core/internal/gc/impl/conservative/gc.d:2830]: Is this related to this commit (use clone() instead of fork()) or a different bug fix or similar?

**Review by llucax — APPROVED** [review]
Besides the very minimal comments, LGTM! :+1:

File: src/core/internal/gc/impl/conservative/gc.d (hunk 1 above)
Comment (rainers) [reply src/core/internal/gc/impl/conservative/gc.d:1393]: Thanks @llucax for review. Indeed this crept in during debugging, but I think it must have been inadvertently been removed at some time as the variables are still declared.

File: src/core/internal/gc/impl/conservative/gc.d (hunk 2 above)
Comment (rainers) [reply src/core/internal/gc/impl/conservative/gc.d:2710]: I'm not so sure either, but it is the version that is also tested for FreeBSD and that passes, too. The deadlock might not manifest on all platforms and probably also depends on the actual C library used.

`_Fork` is just being released in libc, so it might take some time to be available on all platforms.

File: src/core/internal/gc/impl/conservative/gc.d (hunk 3 above)
Comment (rainers) [reply src/core/internal/gc/impl/conservative/gc.d:2830]: It's unrelated. Just the observation that it doesn't help to fork another process just to wait for it to return. We can just do it's work directly.

**Comment by rainers** [comment]
@llucax One other issue that we also discussed in the other PR IIRC: the second commit https://github.com/dlang/druntime/pull/3514/commits/2a80690bb56599a59998ab936a9446bb7bc377f8 switches to atomic operations on the mark bitmaps to avoid corruption caused by concurrent access when allocating.

**Comment by llucax** [comment]
Yeah, I guess it makes sense, but for some reason we never had issues with that using single threaded applications. Maybe it was just luck...

**Review by RazvanN7 — APPROVED** [review]

**Comment by FraMecca** [comment]
> @FraMecca Hi. In an attempt to resurrect your efforts I rebased your PR and squashed your commits here. I hope that is ok with you. Should I rather push to your PR #2604 instead?

No, this is perfectly fine.

**Comment by rainers** [comment]
After tests have passed for another week, I have now disabled forking by default. The GC unittests are now also run with the forking option in addition to the notorious hospital.d which revealed a couple of issues before.

**Comment by RazvanN7** [comment]
Everyone OK with merging this?

**Comment by rainers** [comment]
It now needs a rebase after merging #3523 though...

**Comment by rainers** [comment]
> It now needs a rebase after merging #3523 though...

Done.

**Comment by llucax** [comment]
Amazing!!! Thanks @rainers and @FraMecca for doing this! This gives closure to a piece of work I did with a lot of love and dedication, already 11 years ago and makes it finally available for a wider audience. I really hope it is useful :)

**Comment by rainers** [comment]
Thanks to @llucax and @FraMecca for doing most of the work here.

The precise GC took only about 7 years to land in master, but I'm also not certain how much it is actually used outside of Visual D ;-)
